File: PrimaMateria/PrimaMateria/Classes/XTRGraphViewController.m
//
//  XTRGraphViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 11/21/11.
//  Copyright 2016 xTrensa. All rights reserved.
//

#import "PrimaMateria.h"

static NSString *ATTRIBUTE_NAME   = @"attributeName";
static NSString *TITLE            = @"title";
static NSString *MAJOR_TICK_MARKS = @"majorTickMarks";
static NSString *MINOR_TICK_MARKS = @"minorTickMarks";
static NSString *MAXIMUM_VALUE    = @"maximumValue";
static NSString *MINIMUM_VALUE    = @"minimumValue";

@interface XTRGraphViewController ()
- (NSData *) dataForResource: (NSString *) aResourceName type: (NSString *) aType directory: (NSString *) aDirectory;
- (NSNumber *) element: (XTRElement *) anElement valueForIdentifier: (NSString *) anIdentifier;
- (void) creatBarChart;
- (void) createXAxis: (CPTXYAxisSet *) axisSet majorTickStyle: (CPTLineStyle *) majorTickStyle minorTickStyle: (CPTLineStyle *) minorTickStyle;
- (void) createYAxis: (CPTXYAxisSet *) axisSet minorTicks: (float) minorTicks majorTickStyle: (CPTLineStyle *) majorTickStyle minorTickStyle: (CPTLineStyle *) minorTickStyle majorTicks: (float) majorTicks dict: (NSDictionary *) dict maxValue: (float) maxValue minValue: (float) minValue;
- (void) graphSelected: (NSNotification *) notification;
- (void) showGraphForChoiceAtIndex: (int) anIndex;
@end

@implementation XTRGraphViewController
@synthesize hostingView;
@synthesize barChart;
@synthesize errorString;
@synthesize button;

#pragma mark Private Methods

- (void) creatBarChart {
    barChart = [[CPTXYGraph alloc] initWithFrame: CGRectZero];
    [barChart applyTheme: [CPTTheme themeNamed: kCPTSlateTheme]];
    
    hostingView.hostedGraph = barChart;
    
    barChart.plotAreaFrame.masksToBorder = NO;
    barChart.paddingLeft = 100.0;
    barChart.paddingTop = 10.0;
    barChart.paddingRight = 0.0;
    barChart.paddingBottom = 80.0;
    barChart.delegate = self;
}

- (void) createXAxis: (CPTXYAxisSet *) axisSet majorTickStyle: (CPTLineStyle *) majorTickStyle minorTickStyle: (CPTLineStyle *) minorTickStyle {
    CPTXYAxis *x = axisSet.xAxis;
    x.axisLineStyle = nil;
    x.minorTicksPerInterval = 10;
    x.majorTickLineStyle = majorTickStyle;
    x.minorTickLineStyle = minorTickStyle;
    x.majorIntervalLength = @1;
    x.majorGridLineStyle = minorTickStyle;
    x.orthogonalPosition = @0;
    x.title = @"Atomic Number";
    x.titleLocation = @([XTRDataSource sharedInstance].elementCount / 2);
    x.titleOffset = 35.0f;
    
    // Define some custom labels for the data elements
    x.labelRotation = M_PI / 4;
    x.labelingPolicy = CPTAxisLabelingPolicyNone;
    NSArray *customTickLocations = @[[NSDecimalNumber numberWithInt: 1],
                                    [NSDecimalNumber numberWithInt: 10],
                                    [NSDecimalNumber numberWithInt: 20],
                                    [NSDecimalNumber numberWithInt: 30],
                                    [NSDecimalNumber numberWithInt: 40],
                                    [NSDecimalNumber numberWithInt: 50],
                                    [NSDecimalNumber numberWithInt: 60],
                                    [NSDecimalNumber numberWithInt: 70],
                                    [NSDecimalNumber numberWithInt: 80],
                                    [NSDecimalNumber numberWithInt: 90],
                                    [NSDecimalNumber numberWithInt: 100],
                                    [NSDecimalNumber numberWithInt: 110]];
    NSArray *xAxisLabels = @[ @"1", @"10", @"20", @"30", @"40", @"50", @"60", @"70", @"80", @"90", @"100", @"110"];
    NSUInteger labelLocation = 0;
    NSMutableArray *customLabels = [NSMutableArray arrayWithCapacity:xAxisLabels.count];
    for (NSNumber *tickLocation in customTickLocations) {
        CPTAxisLabel *newLabel = [[CPTAxisLabel alloc] initWithText:xAxisLabels[labelLocation++] textStyle: x.labelTextStyle];
        newLabel.tickLocation = tickLocation;
        newLabel.offset = x.labelOffset + x.majorTickLength;
        newLabel.rotation = M_PI / 4;
        [customLabels addObject: newLabel];
    }
    x.axisLabels =  [NSSet setWithArray: customLabels];
}

- (void) createYAxis: (CPTXYAxisSet *) axisSet minorTicks: (float) minorTicks majorTickStyle: (CPTLineStyle *) majorTickStyle minorTickStyle: (CPTLineStyle *) minorTickStyle majorTicks: (float) majorTicks dict: (NSDictionary *) dict maxValue: (float) maxValue minValue: (float) minValue {
    CPTXYAxis *y = axisSet.yAxis;
    y.axisLineStyle = nil;
    y.minorTicksPerInterval = minorTicks;
    y.majorTickLineStyle = majorTickStyle;
    y.minorTickLineStyle = minorTickStyle;
    y.majorIntervalLength = @(majorTicks);
    y.majorGridLineStyle = minorTickStyle;
    y.orthogonalPosition = @0;
    y.title = dict[TITLE];
    y.titleOffset = 75.0f;
    y.titleLocation = @(maxValue / 2);
    y.labelingOrigin = @(minValue);
}

- (NSNumber *) element: (XTRElement *) anElement valueForIdentifier: (NSString *) anIdentifier {
    NSNumber *aValue = [anElement valueForKey: anIdentifier];
    return (aValue != nil) ? aValue : @0;
}

- (void) showElementPanelForElementAtIndex: (int) anIndex {
    [XTRPropertiesStore storeViewTitle: self.title];
    [XTRPropertiesStore storeAtomicNumber:@(anIndex)];
    [self performSegueWithIdentifier: SHOW_INSPECTOR_FROM_GRAPH_VIEW sender: self];
}

- (void) showGraphForChoiceAtIndex: (int) anIndex {
    NSDictionary *dict = [XTRDataSource sharedInstance].graphPropertyList[anIndex];
    float minValue = [dict[MINIMUM_VALUE] floatValue];
    float maxValue = [dict[MAXIMUM_VALUE] floatValue];
    float majorTicks = [dict[MAJOR_TICK_MARKS] floatValue];
    float minorTicks = [dict[MINOR_TICK_MARKS] floatValue];
    [self creatBarChart];
    CPTXYPlotSpace *plotSpace = (CPTXYPlotSpace *)barChart.defaultPlotSpace;
    plotSpace.yRange = [CPTPlotRange plotRangeWithLocation: @(minValue) length: @(maxValue)];
    plotSpace.xRange = [CPTPlotRange plotRangeWithLocation: @(0.0) length: @([XTRDataSource sharedInstance].elementCount + 1)];
    CPTMutableLineStyle *majorTickStyle = [CPTMutableLineStyle lineStyle];
    majorTickStyle.lineWidth = 2.0f;
    majorTickStyle.lineColor = [CPTColor greenColor];
    
    CPTMutableLineStyle *minorTickStyle = [CPTMutableLineStyle lineStyle];
    minorTickStyle.lineWidth = 1.0f;
    minorTickStyle.lineColor = [CPTColor greenColor];
    
    CPTXYAxisSet *axisSet = (CPTXYAxisSet *)barChart.axisSet;
    [self createXAxis: axisSet majorTickStyle: majorTickStyle minorTickStyle: minorTickStyle];
    [self createYAxis: axisSet minorTicks: minorTicks majorTickStyle: majorTickStyle minorTickStyle: minorTickStyle majorTicks: majorTicks dict: dict maxValue: maxValue minValue: minValue];
    
    CPTBarPlot *barPlot = [CPTBarPlot tubularBarPlotWithColor:[CPTColor whiteColor] horizontalBars: NO];
    barPlot.delegate = self;
    barPlot.barWidth = @1.0f;
    barPlot.baseValue = @0;
    barPlot.dataSource = self;
    barPlot.barOffset = @0.0f;
    barPlot.identifier = dict[ATTRIBUTE_NAME];
    [barChart addPlot: barPlot toPlotSpace: plotSpace];
}

- (void) graphSelected: (NSNotification *) notification {
    NSNumber *object = notification.object;
    [self dismissViewControllerAnimated: YES completion: nil];
    [self showGraphForChoiceAtIndex:object.intValue];
}

- (NSData *) dataForResource: (NSString *) aResourceName type: (NSString *) aType directory: (NSString *) aDirectory {
    return [NSData dataWithContentsOfFile:[[NSBundle bundleForClass:[self class]] pathForResource: aResourceName ofType: aType inDirectory: aDirectory]];
}

#pragma mark - Action Methods

- (IBAction) toolbarItemTapped: (id) sender {
    UIStoryboard *storyboard = [XTRAppDelegate storyboard];
    XTRGraphChoiceViewController *content = [storyboard instantiateViewControllerWithIdentifier: NSStringFromClass([XTRGraphChoiceViewController class])];
    CGSize contentSize = CGSizeMake(294, 664);
    content.modalPresentationStyle = UIModalPresentationPopover;
    content.popoverPresentationController.barButtonItem = sender;
    content.preferredContentSize = contentSize;
   [self presentViewController:content animated:YES completion:nil];
}

#pragma mark - Plot Data Source Methods

- (NSUInteger) numberOfRecordsForPlot: (CPTPlot *) plot {
    return [XTRDataSource sharedInstance].elementCount;
}

- (NSNumber *) numberForPlot: (CPTPlot *) plot field: (NSUInteger) fieldEnum recordIndex: (NSUInteger) index  {
    NSNumber *num = nil;
    if ([plot isKindOfClass:[CPTBarPlot class]]) {
        XTRElement *element = [[XTRDataSource sharedInstance] elementAtIndex: index];

        NSString *identifier = (NSString *)plot.identifier;
        switch (fieldEnum) {
            case 0:
                num = @((int)index + 1);
                break;
                
            case 1:
                if ([identifier isEqualToString: ELEMENT_ATOMIC_MASS])
                    num = element.atomicMass;
                else if ([identifier isEqualToString: ELEMENT_ATOMIC_RADIUS])
                    num = element.atomicRadius;
                else if ([identifier isEqualToString: ELEMENT_BOILING_POINT])
                    num = element.boilingPoint;
                else if ([identifier isEqualToString: ELEMENT_COEFFICIENT_OF_LINEAL_THERMAL_EXPANSION])
                    num = element.coefficientOfLinealThermalExpansionScaled;
                else if ([identifier isEqualToString: ELEMENT_COVALENT_RADIUS])
                    num = element.covalentRadius;
                else if ([identifier isEqualToString: ELEMENT_CROSS_SECTION])
                    num = element.crossSection;
                else if ([identifier isEqualToString: ELEMENT_DENSITY])
                    num = [element valueForKey: ELEMENT_DENSITY];
                else if ([identifier isEqualToString: ELEMENT_ELASTIC_MODULUS_BULK])
                    num = element.elasticModulusBulk;
                else if ([identifier isEqualToString: ELEMENT_ELASTIC_MODULUS_RIGIDITY])
                    num = element.elasticModulusRigidity;
                else if ([identifier isEqualToString: ELEMENT_ELASTIC_MODULUS_YOUNGS])
                    num = element.elasticModulusYoungs;
                else if ([identifier isEqualToString: ELEMENT_ELECTRO_CHEMICAL_EQUIVALENT])
                    num = element.electroChemicalEquivalent;
                else if ([identifier isEqualToString: ELEMENT_ELECTRO_NEGATIVITY])
                    num = element.electroNegativity;
                else if ([identifier isEqualToString: ELEMENT_ELECTRON_WORK_FUNCTION])
                    num = element.electronWorkFunction;
                else if ([identifier isEqualToString: ELEMENT_MELTING_POINT])
                    num = element.meltingPoint;
                else if ([identifier isEqualToString: ELEMENT_ENTHALPY_OF_ATOMIZATION])
                    num = element.enthalpyOfAutomization;
                else if ([identifier isEqualToString: ELEMENT_ENTHALPY_OF_FUSION])
                    num = element.enthalpyOfFusion;
                else if ([identifier isEqualToString: ELEMENT_ENTHALPY_OF_VAPORIZATION])
                    num = element.enthalpyOfVaporization;
                else if ([identifier isEqualToString: ELEMENT_IONIC_RADIUS])
                    num = element.ionicRadius;
                else if ([identifier isEqualToString: ELEMENT_HARDNESS_SCALE_BRINELL])
                    num = element.hardnessScaleBrinell;
                else if ([identifier isEqualToString: ELEMENT_HARDNESS_SCALE_MOHS])
                    num = element.hardnessScaleMohs;
                else if ([identifier isEqualToString: ELEMENT_HARDNESS_SCALE_VICKERS])
                    num = element.hardnessScaleVickers;
                else if ([identifier isEqualToString: ELEMENT_HEAT_OF_FUSION])
                    num = element.heatOfFusion;
                else if ([identifier isEqualToString: ELEMENT_HEAT_OF_VAPORIZATION])
                    num = element.heatOfVaporization;
                else if ([identifier isEqualToString: ELEMENT_IONIZATION_POTENTIAL_FIRST])
                    num = element.ionizationPotentialFirst;
                else if ([identifier isEqualToString: ELEMENT_IONIZATION_POTENTIAL_SECOND])
                    num = element.ionizationPotentialSecond;
                else if ([identifier isEqualToString: ELEMENT_IONIZATION_POTENTIAL_THIRD])
                    num = element.ionizationPotentialThird;
                else if ([identifier isEqualToString: ELEMENT_MOLAR_HEAT_CAPACITY])
                    num = element.molarHeatCapacity;
                else if ([identifier isEqualToString: ELEMENT_MOLAR_VOLUME])
                    num = element.molarVolume;
                else if ([identifier isEqualToString: ELEMENT_SPECIFIC_HEAT_CAPACITY])
                    num = element.specificHeatCapacity;
                else if ([identifier isEqualToString: ELEMENT_VALENCE_ELECTRON_POTENTIAL])
                    num = element.valenceElectronPotential;
                else
                    num = @0;
                break;
        }
    }
    return num;
}

-(void)barPlot:(CPTBarPlot *)plot barWasSelectedAtRecordIndex:(NSUInteger)index {
    [self showElementPanelForElementAtIndex:index];
}

- (CPTFill *) barFillForBarPlot: (CPTBarPlot *) barPlot recordIndex: (NSUInteger) index {
    XTRElement *element = [[XTRDataSource sharedInstance] elementAtIndex: index];
    return [CPTFill fillWithColor:[CPTColor colorWithCGColor:element.seriesColor.CGColor]];
}

#pragma mark - View Management Methods

- (void) viewDidLoad {
    [super viewDidLoad];
    [[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(graphSelected:) name: NOTIFICATION_GRAPH_SELECTED object: nil];
    self.title = @"Graphs";
    [self showGraphForChoiceAtIndex: 0];
}

- (BOOL) shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation) interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

#pragma mark - Memory Management Methods

- (void) dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver: self];
    button = nil;
    hostingView = nil;
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRGraphViewController.swift
//
//  XTRGraphViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/22/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

struct GraphAttribute {
    static let name = "attributeName"
    static let title = "title"
    static let majorTickMarks = "majorTickMarks"
    static let minorTickMarks = "minorTickMarks"
    static let maximumValue = "maximumValue"
    static let minimumValue = "minimumValue"
}

class XTRGraphViewController : UIViewController, CPTPlotDataSource, CPTBarPlotDelegate {
    
    @IBOutlet var hostingView : CPTGraphHostingView!
    @IBOutlet var button : UIBarButtonItem!
    
    var barChart : CPTXYGraph?
    var errorString : String?
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if segue.identifier == "showGraphChoice" {
            let dest = segue.destination
            let aFrame = CGRect(x: 0, y: 0, width: 298, height: 668)

            dest.view.frame = aFrame
            dest.preferredContentSize = CGSize(width: 298, height: 668)
        }
    }

    // MARK: - Internal Methods
    
    func creatBarChart() {
        barChart = CPTXYGraph(frame: CGRect.zero)
        barChart!.apply(CPTTheme(named: CPTThemeName.slateTheme))
        
        hostingView.hostedGraph = barChart
        
        barChart!.plotAreaFrame!.masksToBorder = false
        barChart!.paddingLeft = 100.0
        barChart!.paddingTop = 10.0
        barChart!.paddingRight = 0.0
        barChart!.paddingBottom = 80.0
        barChart!.delegate = self
    }
    
    func createXAxis(_ axisSet :CPTXYAxisSet, majorTickStyle: CPTLineStyle,  minorTickStyle: CPTLineStyle) {
        let x = axisSet.xAxis!
        let customTickLocations = [1, 10, 20, 30,  40, 50, 60, 70, 80, 90, 100, 110]
        let xAxisLabels = ["1", "10", "20", "30", "40", "50", "60", "70", "80", "90", "100", "110"]
        var customLabels : [CPTAxisLabel] = []
        let textStyle = CPTMutableTextStyle.init()
        
        textStyle.color = CPTColor.black()
        textStyle.fontName = "Verdana-Bold"
        textStyle.fontSize = 18.0
        textStyle.textAlignment = CPTTextAlignment.center
        
        x.axisLineStyle = nil
        x.minorTicksPerInterval = 10
        x.majorTickLineStyle = majorTickStyle
        x.minorTickLineStyle = minorTickStyle
        x.majorIntervalLength = 1
        x.majorGridLineStyle = minorTickStyle
        x.orthogonalPosition = 0
        x.title = "Atomic Number"
        x.titleLocation = XTRDataSource.sharedInstance().elementCount() / 2 as NSNumber?
        x.titleOffset = 30.0
        x.titleTextStyle = textStyle
        
        // Define some custom labels for the data elements
        x.labelRotation = CGFloat.pi / 4.0
        x.labelingPolicy = CPTAxisLabelingPolicy.none
        
        for index in 0..<customTickLocations.count {
            let tickLocation = customTickLocations[index]
            let labelValue = xAxisLabels[index]
            let newLabel = CPTAxisLabel.init(text: labelValue, textStyle: x.labelTextStyle)
            newLabel.tickLocation = NSNumber(value: tickLocation)
            newLabel.offset = x.labelOffset + x.majorTickLength
            newLabel.rotation = CGFloat.pi / 4.0
            customLabels.append(newLabel)
        }
        
        let aSet = NSSet(array: customLabels as [AnyObject])
        x.axisLabels =  aSet as? Set<CPTAxisLabel>
    }
    
    func createYAxis(_ axisSet: CPTXYAxisSet, minorTicks: Float, majorTickStyle: CPTLineStyle, minorTickStyle: CPTLineStyle, majorTicks: Float, dict: [String : AnyObject], maxValue: Float, minValue: Float) {
        let y = axisSet.yAxis!
        let title = dict[GraphAttribute.title] as? String
        let textStyle = CPTMutableTextStyle.init()
        
        textStyle.color = CPTColor.black()
        textStyle.fontName = "Verdana-Bold"
        textStyle.fontSize = 18.0
        textStyle.textAlignment = CPTTextAlignment.center
        
        y.axisLineStyle = nil
        y.minorTicksPerInterval = UInt(minorTicks)
        y.majorTickLineStyle = majorTickStyle
        y.minorTickLineStyle = minorTickStyle
        y.majorIntervalLength = NSNumber(value: majorTicks)
        y.majorGridLineStyle = minorTickStyle
        y.orthogonalPosition = 0
        y.title = title
        y.titleOffset = 65.0
        y.titleLocation = NSNumber(value: maxValue / 2.0)
        y.labelingOrigin = NSNumber(value: minValue)
        y.titleTextStyle = textStyle
    }
    
    func element(_ anElement: XTRElement, anIdentifier: String) -> NSNumber {
        let aValue : NSNumber? = anElement.value(forKeyPath: anIdentifier) as? NSNumber
        return (aValue != nil) ? aValue! : 0
    }
    
    func showElementPanelForElementAtIndex(_ anIndex: Int) {
        XTRPropertiesStore.viewTitle = title!
        XTRPropertiesStore.atomicNumber = anIndex
        performSegue(withIdentifier: SegueName.showInspectorFromGraphView, sender: self)
    }
    
    func showGraphForChoiceAtIndex(_ anIndex: UInt) {
        let dict = XTRDataSource.sharedInstance().graphPropertyList?[Int(anIndex)]
        let minValue = dict?[GraphAttribute.minimumValue] as! Float
        let maxValue = dict?[GraphAttribute.maximumValue] as! Float
        let majorTicks = dict?[GraphAttribute.majorTickMarks] as! Float
        let minorTicks = dict?[GraphAttribute.minorTickMarks] as! Float
        let majorTickStyle = CPTMutableLineStyle.init()
        let minorTickStyle = CPTMutableLineStyle.init()
        let length = XTRDataSource.sharedInstance().elementCount() + 1
        
        creatBarChart()
        
        let plotSpace = barChart!.defaultPlotSpace as! CPTXYPlotSpace
        let axisSet = barChart!.axisSet as! CPTXYAxisSet
        let barPlot = XTRBarPlot.tubularBarPlot(with: CPTColor.white(), horizontalBars: false)
        
        plotSpace.yRange = CPTPlotRange.init(location: NSNumber(value: Float(minValue)), length: NSNumber(value: Float(maxValue)))
        plotSpace.xRange = CPTPlotRange.init(location: 0.0, length: NSNumber(value: length))
        majorTickStyle.lineWidth = 2.0
        majorTickStyle.lineColor = CPTColor.darkGray()
        
        minorTickStyle.lineWidth = 1.0
        minorTickStyle.lineColor = CPTColor.darkGray()
        
        createXAxis(axisSet, majorTickStyle: majorTickStyle, minorTickStyle: minorTickStyle)
        createYAxis(axisSet, minorTicks: minorTicks, majorTickStyle: majorTickStyle, minorTickStyle: minorTickStyle, majorTicks: majorTicks, dict: dict!, maxValue: maxValue, minValue: minValue)
        
        barPlot.element = XTRDataSource.sharedInstance().elementAtIndex(Int(anIndex))
        barPlot.delegate = self
        barPlot.barWidth = 1.0
        barPlot.baseValue = 0
        barPlot.dataSource = self
        barPlot.barOffset = 0.0
        barPlot.identifier = dict?[GraphAttribute.name] as! String as (NSCoding & NSCopying & NSObjectProtocol)?
        barPlot.barCornerRadius = 0.0
        
        barChart?.add(barPlot, to: plotSpace)
    }
    
    @objc func dataForResource(_ aResourceName: String, type: String, directory: String) -> Data {
        return try! Data(contentsOf: URL(fileURLWithPath: Bundle(for: classForCoder).path(forResource: aResourceName, ofType: type, inDirectory: directory)!))
    }
    
    // MARK: - Action Methods
        
    // MARK: - Notification Methods
    
    @objc func graphSelected(_ notification: Notification) {
        let object = notification.object as! NSNumber
        
        dismiss(animated: true, completion: nil)
        showGraphForChoiceAtIndex(UInt(object.intValue))
    }
    
    // MARK: - Plot Data Source Methods
    
    @objc func numberOfRecords(for plot: CPTPlot) -> UInt {
        return UInt(XTRDataSource.sharedInstance().elementCount())
    }
    
    @objc func number(for plot: CPTPlot, field fieldEnum: UInt, record idx: UInt) -> Any? {
        var num : NSNumber = 0
        
        if plot.isKind(of: CPTBarPlot.classForCoder()) {
            let element = XTRDataSource.sharedInstance().elementAtIndex(Int(idx))
            let identifier = plot.identifier as! String
            
            switch (fieldEnum) {
            case 0:
                num = NSNumber(value: idx + 1)
            case 1:
                if identifier == ELEMENT_ATOMIC_MASS {
                    num = NSNumber(value: element.atomicMass)
                } else if identifier == ELEMENT_ATOMIC_RADIUS {
                    num = NSNumber(value: element.atomicRadius)
                } else if identifier == ELEMENT_BOILING_POINT {
                    num = NSNumber(value: element.boilingPoint)
                } else if identifier == ELEMENT_COEFFICIENT_OF_LINEAL_THERMAL_EXPANSION {
                    num = NSNumber(value: element.coefficientOfLinealThermalExpansionScaled)
                } else if identifier == ELEMENT_COVALENT_RADIUS {
                    num = NSNumber(value: element.covalentRadius)
                } else if identifier == ELEMENT_CROSS_SECTION {
                    num = NSNumber(value: element.crossSection)
                } else if identifier == ELEMENT_DENSITY {
                    num = NSNumber(value: element.density)
                } else if identifier == ELEMENT_ELASTIC_MODULUS_BULK {
                    num = NSNumber(value: element.elasticModulusBulk)
                } else if identifier == ELEMENT_ELASTIC_MODULUS_RIGIDITY {
                    num = NSNumber(value: element.elasticModulusRigidity)
                } else if identifier == ELEMENT_ELASTIC_MODULUS_YOUNGS {
                    num = NSNumber(value: element.elasticModulusYoungs)
                } else if identifier == ELEMENT_ELECTRO_CHEMICAL_EQUIVALENT {
                    num = NSNumber(value: element.electroChemicalEquivalent)
                } else if identifier == ELEMENT_ELECTRO_NEGATIVITY {
                    num = NSNumber(value: element.electroNegativity)
                } else if identifier == ELEMENT_ELECTRON_WORK_FUNCTION {
                    num = NSNumber(value: element.electronWorkFunction)
                } else if identifier == ELEMENT_MELTING_POINT {
                    num = NSNumber(value: element.meltingPoint)
                } else if identifier == ELEMENT_ENTHALPY_OF_ATOMIZATION {
                    num = NSNumber(value: element.enthalpyOfAutomization)
                } else if identifier == ELEMENT_ENTHALPY_OF_FUSION {
                    num = NSNumber(value: element.enthalpyOfFusion)
                } else if identifier == ELEMENT_ENTHALPY_OF_VAPORIZATION {
                    num = NSNumber(value: element.enthalpyOfVaporization)
                } else if identifier == ELEMENT_IONIC_RADIUS {
                    num = NSNumber(value: element.ionicRadius)
                } else if identifier == ELEMENT_HARDNESS_SCALE_BRINELL {
                    num = NSNumber(value: element.hardnessScaleBrinell)
                } else if identifier == ELEMENT_HARDNESS_SCALE_MOHS {
                    num = NSNumber(value: element.hardnessScaleMohs)
                } else if identifier == ELEMENT_HARDNESS_SCALE_VICKERS {
                    num = NSNumber(value: element.hardnessScaleVickers)
                } else if identifier == ELEMENT_HEAT_OF_FUSION {
                    num = NSNumber(value: element.heatOfFusion)
                } else if identifier == ELEMENT_HEAT_OF_VAPORIZATION {
                    num = NSNumber(value: element.heatOfVaporization)
                } else if identifier == ELEMENT_IONIZATION_POTENTIAL_FIRST {
                    num = NSNumber(value: element.ionizationPotentialFirst)
                } else if identifier == ELEMENT_IONIZATION_POTENTIAL_SECOND {
                    num = NSNumber(value: element.ionizationPotentialSecond)
                } else if identifier == ELEMENT_IONIZATION_POTENTIAL_THIRD {
                    num = NSNumber(value: element.ionizationPotentialThird)
                } else if identifier == ELEMENT_MOLAR_HEAT_CAPACITY {
                    num = NSNumber(value: element.molarHeatCapacity)
                } else if identifier == ELEMENT_MOLAR_VOLUME {
                    num = NSNumber(value: element.molarVolume)
                } else if identifier == ELEMENT_SPECIFIC_HEAT_CAPACITY {
                    num = NSNumber(value: element.specificHeatCapacity)
                } else if identifier == ELEMENT_VALENCE_ELECTRON_POTENTIAL {
                    num = NSNumber(value: element.valenceElectronPotential)
                } else {
                    num = 0
                }
            default:
                num = 0
                break
            }
        }
        
        return num
    }
    
    @objc func barPlot(_ plot: CPTBarPlot, barWasSelectedAtRecord index: UInt) {
        showElementPanelForElementAtIndex(Int(index))
    }
    
    @objc func barFillForBarPlot(_ barPlot: CPTBarPlot, recordIndex: UInt) -> CPTFill {
        let element = XTRDataSource.sharedInstance().elementAtIndex(Int(recordIndex))
        return CPTFill(color: CPTColor(cgColor: element.seriesColor.cgColor))
    }
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        NotificationCenter.default.addObserver(self, selector: #selector(XTRGraphViewController.graphSelected(_:)), name: .graphSelectedNotification, object: nil)
        
        showGraphForChoiceAtIndex(0)
        if #available(iOS 11.0, *) {
            navigationController?.navigationBar.prefersLargeTitles = true
        } else {
            // Fallback on earlier versions
        }
    }
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        NotificationCenter.default.removeObserver(self, name: .graphSelectedNotification, object: nil)
        button = nil
        hostingView = nil
    }
    
}
